pkg/controller/core/workload_controller.go

```diff
@@ -154,14 +154,15 @@ func (r *WorkloadReconciler) Delete(e event.DeleteEvent) bool {
 	// the state is unknown, the workload could have been assumed and we need
 	// to clear it from the cache.
 	if wl.Spec.Admission != nil || e.DeleteStateUnknown {
-		if err := r.cache.DeleteWorkload(wl); err != nil {
-			if !e.DeleteStateUnknown {
-				log.Error(err, "Failed to delete workload from cache")
-			}
-		}
 
 		// trigger the move of associated inadmissibleWorkloads if required.
-		r.queues.QueueAssociatedInadmissibleWorkloads(ctx, wl)
+		r.queues.QueueAssociatedInadmissibleWorkloadsAfter(ctx, wl, func() {
+			if err := r.cache.DeleteWorkload(wl); err != nil {
```

Comment:
You may want to add a comment here on why we are doing this: The workload is deleted from the cache while holding the queue lock to ensure that potentially high priority workloads are requeued first, and then the deleted workload's resources are released. This prevents a lower priority workload waiting in the queue from obtaining the released resources before the high priority ones.